server/src/main/java/com/cloud/storage/StorageManagerImpl.java

```diff
@@ -1148,6 +1151,95 @@ public PrimaryDataStoreInfo updateStoragePool(UpdateStoragePoolCmd cmd) throws I
+            ClusterVO clusterVO = _clusterDao.findById(clusterId);
+            hypervisorType = clusterVO.getHypervisorType();
+        } else {
+            hypervisorType = primaryStorage.getHypervisor();
+        }
+        if (!zoneWidePoolSupportedHypervisorTypes.contains(hypervisorType)) {
+            throw new InvalidParameterValueException("Primary storage scope change is not supported for hypervisor type " + hypervisorType);
+        }
+
+        String providerName = primaryStorage.getStorageProviderName();
+        DataStoreProvider storeProvider = _dataStoreProviderMgr.getDataStoreProvider(providerName);
+        PrimaryDataStoreLifeCycle lifeCycle = (PrimaryDataStoreLifeCycle) storeProvider.getDataStoreLifeCycle();
+        DataStore primaryStore = _dataStoreMgr.getPrimaryDataStore(id);
+
+        Long zoneId = primaryStorage.getDataCenterId();
+        DataCenterVO zone = _dcDao.findById(zoneId);
+        if (zone == null) {
+            throw new InvalidParameterValueException("Unable to find zone by id " + zoneId);
+        }
+        if (Grouping.AllocationState.Disabled == zone.getAllocationState()) {
+            throw new PermissionDeniedException("Cannot perform this operation, Zone is currently disabled: " + zoneId);
+        }
+
+        if (newScope.equals(ScopeType.ZONE)) {
+            ClusterScope clusterScope = new ClusterScope(primaryStorage.getClusterId(), null, zoneId);
+            lifeCycle.changeStoragePoolScopeToZone(primaryStore, clusterScope, hypervisorType);
+
+        } else {
+            Long clusterId = cmd.getClusterId();
+            if (clusterId == null) {
+                throw new IllegalArgumentException("Unable to find cluster with ID: " + clusterId);
+            }
+            ClusterVO clusterVO = _clusterDao.findById(clusterId);
+            if (clusterVO == null) {
+                throw new InvalidParameterValueException("Unable to find cluster by id " + clusterId);
+            }
+            if (Grouping.AllocationState.Disabled == clusterVO.getAllocationState()) {
+                throw new PermissionDeniedException("Cannot perform this operation, Cluster is currently disabled: " + zoneId);
+            }
+            List<VirtualMachine.State> states = Arrays.asList(State.Starting, State.Running, State.Stopping, State.Migrating, State.Restoring);
+            List<VolumeVO> volumesInOtherClusters = volumeDao.listByPoolIdVMStatesNotInCluster(clusterId, states, id);
+            if (volumesInOtherClusters.size() > 0) {
+                throw new CloudRuntimeException("Cannot change scope of the pool " + primaryStorage.getName() + " to cluster " + clusterVO.getName() + " as there are associated volumes present for other clusters");
+            }
+
+            ClusterScope clusterScope = new ClusterScope(clusterId, clusterVO.getPodId(), zoneId);
+            lifeCycle.changeStoragePoolScopeToCluster(primaryStore, clusterScope, hypervisorType);
+        }
+
+        return true;
+    }
```

Comment:
I know this method is not that complex and long, but you could break into smaller ones, making the code more readable and add some unit test as well. What do you think?

Reply:
Good idea. Will do this in next commit.